README.md
# thiefringer

ThiefRinger - alarm system.

Threads like microservices - They running simultaneously and communicate via Queue.
Queue like a message bus.

**NOTE!** This is a quick and dirty script for hobby usage only!
* This is not a microservice architecture -> It uses multithreading and messaging queues
* This is not a plugin-based architecture -> It utilizes config sections and external libs

***It is a miniature script; It runs on OnionOmega which is an embedded linux board running OpenWrt OS; It tries to use as less overhead as possible to do the job.***

## Installation

### libs

Install extra libraries that this package depends on:
``` sh
opkg install python-light pyOnionGpio pyOledExp power-dock2
# clone and install https://github.com/kyuzom/gpydem
pip install gpydem
# clone and install https://github.com/kyuzom/ooutils
pip install ooutils
```

### pip

Install this package via **pip** package manager:
``` sh
pip install thiefringer
```

### setup.py

Install this package manually via **setup.py** file:
``` sh
git clone https://github.com/kyuzom/thiefringer
cd thiefringer
python setup.py install
```

## Configuration

Config file is used to determine custom behavior a.k.a. client-side logic.
It is JSON formatted.

Parameters:
``` text
  {
      PIR [map] - PIR motion detector related configurations, default: {}
      {
          pin          [int] - GPIO pin number,                          default: -1
          active_value [str] - GPIO logic value when motion is detected, default: "0"
          alarm        [map] - Alert message related configurations, default: {}
          {
              number  [str] - Phone number, default: ""
              message [str] - SMS Message,  default: "ALERT"
          }
      }
      GSM [map] - GSM 3G USB modem related configurations, default: {}
      {
          modem_type [str]   - GSM modem type Id,                             default: ""
          dev_id     [str]   - GSM modem character device (serial port Id),   default: "/dev/ttyS0"
          baudrate   [int]   - GSM modem BaudRate (serial port speed),        default: 9600
          PIN        [str]   - SIM card PIN code,                             default: ""
          re_timeout [float] - GSM modem timeout to reboot (after PIN setup), default: 1
      }
      Battery [map] - LiPo battery level check, default: {}
      {
          frequency   [float] - Battery level measurement frequency in [s],   default: 60
          timeout     [float] - Battery level measurement timeout in [s],     default: 1
          vmax        [float] - Maximum battery level in [V],                 default: 4.2
          vmin        [float] - Minimum battery level in [V],                 default: 3.5
          vpthreshold [float] - Threshold percentage of battery level in [%], default: 20
          alarm       [map]   - Battery message related configurations, default: {}
          {
              number  [str] - Phone number, default: ""
              message [str] - SMS Message,  default: "BATTERY"
          }
      }
  }
```

## Usage

### cmd line

Run when your config file *.thiefringer.json* is next to your python script:
``` sh
python /path/to/thiefringer.py
```

Run with specify path manually to your config file:
``` sh
python /path/to/thiefringer.py -c /path/to/.thiefringer.json
```

### daemon

**NOTE!** Additional files can be found under the [init.d](system/etc/init.d) directory.

Prepare **Procd** init system files on an OpenWrt system, then start in the background, then enable it:
``` sh
/etc/init.d/thiefringer start
/etc/init.d/thiefringer enable
```

## Board

Onion Omega2 Pro - board specific notes and guidelines.

### opkg

Edit the repository source file to access additional packages:
``` sh
vi /etc/opkg/distfeeds.conf
```

Guide what to do manually via the default text editor **Vi**:
``` text
i
uncomment -> src/gz openwrt_base http://downloads.openwrt.org/releases/18.06-SNAPSHOT/packages/mipsel_24kc/base
escape
:x
```

Then install any additional package you want:
``` sh
opkg update
opkg install nano mc
```

### pivot-overlay

**Booting from External Storage**

Prepare the system with the necessary additional packages:
``` sh
opkg update
opkg install swap-utils block-mount e2fsprogs
```

Pack the FileSystem into a separate place aka **External Storage**, therefore from now on it will boot from there:
``` sh
tar -C /overlay -cvf - . | tar -C /mnt/mmcblk0p1 -xf -
block detect > /etc/config/fstab
```

Edit the **fstab** file to change the location of your FileSystem:
``` sh
nano /etc/config/fstab
```

Guide what to do manually via your favorite text editor **nano**:
``` text
replace -> option  target  '/mnt/mmcblk0p1' -> option  target  '/overlay'
replace -> option  enabled '0' -> option  enabled '1'
ctrl+x
y
```

Restart the system:
``` sh
reboot
```

### turn off LEDs

Put these lines at the end of your ***/etc/rc.local*** file:
``` sh
# Turn off full-color LED
omega2-ctrl gpiomux set uart2 pwm23
echo -en '\x00\x00\x00' > /dev/ledchain2
```

Make the script executable:
``` sh
chmod +x /etc/rc.local
```

## License

thiefringer is MIT licensed. See the included [LICENSE](LICENSE) file.


CHANGELOG.md
# Change Log

Changes to thiefringer package.

## [0.0.3] - 2021-12-05

**Improvements**

 - LiPo battery level supports termination, frequency

**Changes**

 - Alerts only sent one time after restart

## [0.0.2] - 2021-12-03

**Improvements**

 - SMS sending capability: support 3G GSM modems
 - LiPo battery level [%] checking
 - Thread checking: support heartbeat mechanism

## [0.0.1] - 2021-10-18

**Improvements**

 - Motion detection: support motion detector sensor as GPIO input; selected via config file
 - Multi-threading: support multiple actions simultaneously
 - Signals: handle external OS-level signals (shutdown correctly)


thiefringer/thiefringer.py
#!/usr/bin/env python
'''
ThiefRinger alarm system - Onion Omega specific python package, motion detection, SMS notification.
'''
from __future__ import print_function
import signal
import sys
import os
import argparse
import json
import threading
try:
	from Queue import Queue
except (ImportError, ModuleNotFoundError):
	from queue import Queue
import onionGpio
import gpydem
import ooutils

class ThiefRinger(object):
	'''
	ThiefRinger alarm object.
	'''
	def __init__(self, opt):
		self.opt = opt
		with open(self.opt.config, 'r') as f:
			config = json.load(f)
		self.main_ctl = threading.Event()
		self.thread_ctl = threading.Event()
		self.thread_msgq = Queue(maxsize=2)
		self.threads = []
		# setup OS signal handlers
		signal.signal(signal.SIGINT, self.signal_terminate)
		signal.signal(signal.SIGTERM, self.signal_terminate)
		#signal.signal(signal.SIGKILL, self.signal_terminate)
		# config defaults
		self.cfg_PIR = config.get('PIR', {})
		self.cfg_GSM = config.get('GSM', {})
		self.cfg_Battery = config.get('Battery', {})

	def run(self):
		'''
		Run alarm system main loop - blocking.
		Start alarm system, wait for external signal, stop alarm system.
		'''
		if self.opt.verbose:
			print('run')
		self.stop_threads()
		self.start_threads()
		while not self.main_ctl.wait(60.0):
			pass
		self.stop_threads(wait=True)

	def terminate(self):
		'''
		Terminate alarm system main loop - nonblocking.
		'''
		if self.opt.verbose:
			print('terminate')
		self.main_ctl.set()

	def signal_terminate(self, *args):
		'''
		Catch process exit signals.
		@param *args: [list] List of additional arguments
		'''
		if self.opt.verbose:
			print('signal_terminate')
		self.terminate()

	def start_threads(self):
		'''
		Start alarm system - nonblocking.
		'''
		if self.opt.verbose:
			print('start_threads')
		self.thread_ctl.clear()
		self.thread_msgq = Queue()#self.thread_msgq.clear()
		self.threads = [
			threading.Thread(target=self.PIR_motion,      args=(self.cfg_PIR,     self.thread_ctl, self.thread_msgq), name='PIR_motion'),
			threading.Thread(target=self.GSM_modem,       args=(self.cfg_GSM,     self.thread_ctl, self.thread_msgq), name='GSM_modem'),
			threading.Thread(target=self.Battery_monitor, args=(self.cfg_Battery, self.thread_ctl, self.thread_msgq), name='Battery_monitor')
		]
		self.threads.append(
			threading.Thread(target=self.thread_heartbeat, args=(self.threads, self.thread_ctl))
		)
		if self.opt.verbose:
			print('num of threads: %d' % len(self.threads))
		for t in self.threads:
			t.start()

	def stop_threads(self, wait=False):
		'''
		Stop alarm system - nonblocking.
		@param wait: [bool] Wait until all alarms are sent or not 
		'''
		if self.opt.verbose:
			print('stop_threads')
		self.thread_ctl.set()
		if wait:
			self.thread_msgq.join()
		for t in self.threads:
			t.join()

	def PIR_motion(self, cfg, ctl, msgq):
		'''
		PIR motion detector main loop.
		@param cfg: [dict] Configurations
		@param ctl: [object] Threading control object (Event)
		@param msgq: [object] Message Queue
		'''
		if self.opt.verbose:
			print('PIR_motion')
		# config defaults
		cfg_pin = cfg.get('pin', -1)
		cfg_active_value = cfg.get('active_value', '0')
		cfg_alarm = cfg.get('alarm', {})
		cfg_alarm_number = cfg_alarm.get('number', '')
		cfg_alarm_message = cfg_alarm.get('message', 'ALERT')
		if sys.version_info[0] == 2:
			cfg_alarm_number = str(cfg_alarm_number)
			cfg_alarm_message = str(cfg_alarm_message)
		# PIR motion detector GPIO pin init
		gpio = onionGpio.OnionGpio(cfg_pin)
		if gpio.setInputDirection():
			raise RuntimeError("Could not set GPIO direction to input!")
		prev_value = '0' if cfg_active_value == '1' else '1'
		# infinite loop
		malert = False
		while not ctl.wait(0.1):
			value = gpio.getValue().strip()
			if prev_value != value:
				if (value == cfg_active_value) and (not malert):
					malert = True
					if self.opt.verbose:
						print("PIR_motion ALERT: motion detected!")
					msgq.put((cfg_alarm_number, cfg_alarm_message), block=True, timeout=0.1)
				prev_value = value
		if self.opt.verbose:
			print('PIR_motion exit')

	def GSM_modem(self, cfg, ctl, msgq):
		'''
		GSM 3G USB modem main loop.
		@param cfg: [dict] Configurations
		@param ctl: [object] Threading control object (Event)
		@param msgq: [object] Message Queue
		'''
		if self.opt.verbose:
			print('GSM_modem')
		# config defaults
		cfg_modem_type = cfg.get('modem_type', '')
		cfg_dev_id = cfg.get('dev_id', '/dev/ttyS0')
		cfg_baudrate = cfg.get('baudrate', 9600)
		cfg_PIN = cfg.get('PIN', '')
		cfg_re_timeout = cfg.get('re_timeout', 1)
		# GSM 3G USB modem serial port init
		gsm = gpydem.Modem.get(cfg_modem_type, cfg_dev_id, baudrate=cfg_baudrate, PIN=cfg_PIN, re_timeout=cfg_re_timeout)
		# infinite loop
		while not ctl.wait(1.0):
			if not msgq.empty():
				number, message = msgq.get(block=True, timeout=0.1)
				msgq.task_done()
				if self.opt.verbose:
					print('GSM_modem sending SMS: {0}, {1}'.format(number, message))
				gsm.sendSMS(number, message)
				if self.opt.verbose:
					print('GSM_modem SMS sent.')
		if self.opt.verbose:
			print('GSM_modem exit')

	def Battery_monitor(self, cfg, ctl, msgq):
		'''
		LiPo battery level check main loop.
		@param cfg: [dict] Configurations
		@param ctl: [object] Threading control object (Event)
		@param msgq: [object] Message Queue
		'''
		if self.opt.verbose:
			print('Battery_monitor')
		# config defaults
		cfg_frequency = cfg.get('frequency', 60)
		cfg_timeout = cfg.get('timeout', 1)
		cfg_vmax = cfg.get('vmax', 4.2)
		cfg_vmin = cfg.get('vmin', 3.5)
		cfg_vpthreshold = cfg.get('vpthreshold', 20)
		cfg_alarm = cfg.get('alarm', {})
		cfg_alarm_number = cfg_alarm.get('number', '')
		cfg_alarm_message = cfg_alarm.get('message', 'BATTERY')
		if sys.version_info[0] == 2:
			cfg_alarm_number = str(cfg_alarm_number)
			cfg_alarm_message = str(cfg_alarm_message)
		# infinite loop
		battery = ooutils.ABattery()
		id = -1
		vprevious = -1
		valert = False
		while not ctl.wait(cfg_frequency):
			id = battery.percentage(vmax=cfg_vmax, vmin=cfg_vmin)
			opcode, vactual = battery.wait(id, timeout=cfg_timeout)
			if self.opt.verbose:
				print('Battery_monitor opcode: {0}, result: {1}'.format(opcode, vactual))
			battery.terminate(id)
			if opcode != ooutils.ABattery.OP_SUCCESS:
				continue
			if self.opt.verbose:
				print('Battery_monitor level: {0} %'.format(vactual))
			if vprevious != vactual:
				vprevious = vactual
				if vactual <= cfg_vpthreshold:
					if self.opt.verbose:
						print('Battery_monitor level low!')
					if not valert:
						valert = True	# send notification only once
						msgq.put((cfg_alarm_number, cfg_alarm_message), block=True, timeout=0.1)
						#return			# stop further measurement
		battery.terminate(id)
		if self.opt.verbose:
			print('Battery_monitor exit')

	def thread_heartbeat(self, threads, ctl):
		'''
		Thread heartbeat main loop.
		@param threads: [list] Threads to check periodically
		@param ctl: [object] Threading control object (Event)
		'''
		if self.opt.verbose:
			print('thread_heartbeat')
		# infinite loop
		while not ctl.wait(1.0):
			for t in threads:
				if not t.is_alive():
					print('Thread {0} is dead! See previous messages for more details.'.format(t.name), file=sys.stderr)
					self.terminate()
		if self.opt.verbose:
			print('thread_heartbeat exit')

def main():
	'''
	Main ThiefRinger function.
	'''
	arg_parser = argparse.ArgumentParser(description='ThiefRinger alarm system.')
	arg_parser.add_argument('-c', '--config',  type=str, default=os.path.join(os.path.abspath(os.path.dirname(__file__)), '.thiefringer.json'), help='Config file')
	arg_parser.add_argument('-v', '--verbose', action='store_true', help='Increase verbosity')
	opt = arg_parser.parse_args()
	try:
		alarm = ThiefRinger(opt)
		rc = alarm.run()
	except KeyboardInterrupt:
		alarm.terminate()
	except Exception as e:
		print(str(e), file=sys.stderr)
		print("\033[91mUnexpected error happened! Please see the details above.\033[0m", file=sys.stderr)
		sys.exit(1)
	else:
		sys.exit(rc)

if __name__ == '__main__':
	main()
